### PR TITLE
MT3-228 #in-progress #comment Create an IndexedDB store

### PR DESCRIPTION
**Pull request by erbridge**
# What?

This creates a new IndexedDB wrapper for creating and interacting with the IndexedDB.

This wrapper depends on two new packages `idb`, a wrapper for IndexedDB to make the interface more usable, and `fake-indexeddb`, for testing.

# Why?

We need to persist the user data to the backend. Rather than building our frontend to be tightly coupled to an API, we want a layer that can sit between the application and the API.

We know that we will need to support offline use of these forms. We also want users to be able to recover unsubmitted forms if they haven't gone online between sessions.

We also know that some forms will require the upload of images.

IndexedDB is a [well supported](https://caniuse.com/#feat=indexeddb) browser database that persists data across sessions. It supports storing blob data (such as images), and has decent storage limits.

# Notes

Node doesn't have a native implementation of IndexedDB, so we install `fake-indexeddb`. It doesn't look like `fake-indexeddb` supports the full IndexedDB spec that `idb` supports, so I have included some comments about the compromizes I've made while implementing it.

# Next steps

- [ ] integrate with an example
- [ ] add JSDoc strings

### CODE REVIEW

File: src/store/Database.ts [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+      autoCloseOnBlocking = false
+    }: OpenOptions<S> = {}
+  ): Promise<Database<NS, N>> {
+    const db = await new Promise<IDBPDatabase<S>>(
+      (resolvePromise, rejectPromise) => {
+        let isRejected = false;
+
+        const resolve = ([db]: [IDBPDatabase<S>, void]): void => {
+          if (isRejected) {
+            db.close();
```
[the remark this answers is not in the document]
Comment (erbridge) [reply src/store/Database.ts]: I will add some comments to (hopefully) address this.

File: src/store/Database.ts [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    {
+      upgrade,
+      blocked,
+      blocking,
+      autoCloseOnBlocking = false
+    }: OpenOptions<S> = {}
+  ): Promise<Database<NS, N>> {
+    const db = await new Promise<IDBPDatabase<S>>(
+      (resolvePromise, rejectPromise) => {
+        let isRejected = false;
```
[the remark this answers is not in the document]
Comment (erbridge) [reply src/store/Database.ts]: I will add some comments to (hopefully) address this.

File: src/store/Database.ts [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+          }
+        };
+
+        const reject = (reason: Error): void => {
+          isRejected = true;
+
+          rejectPromise(reason);
+        };
+
+        const promiseToOpen = openDB<S>(name, version, {
```
[the remark this answers is not in the document]
Comment (erbridge) [reply src/store/Database.ts]: The problem is that the callbacks are outside of the promise lifecycle, so in order to catch them, we need to wrap this promise in its own promise. I hope my incoming comments will help clarify this.

File: src/store/Database.test.ts [hunk 4]
```diff
@@ -0,0 +1,679 @@
...
+            throw new Error(error);
+          }
+        });
+      }).rejects.toThrowError(error);
+    });
+  });
+
+  describe("with a blocked callback", () => {
+    it("calls the provided blocked callback and returns a database instance when attempting to reopen a database that's already open with a different schema version and a blocked callback which resolves the block", async () => {
+      const initialVersion = 5;
```
[the remark this answers is not in the document]
Comment (erbridge) [reply src/store/Database.test.ts]: Do you mean the setup phase of the test, or do you mean the test name?

[the remark this answers is not in the document]
Comment (Floppy) [reply src/store/Database.test.ts]: yeah, setup is the word I was looking for - I think I was running off to do something else when I wrote this :)

[the remark this answers is not in the document]
Comment (erbridge) [reply src/store/Database.test.ts]: :). I can pull that up. I do actually find it more readable to not have to go looking for context outside of a test. Do you have a strong read on this?

**Comment by erbridge** [comment]
@Floppy I think the comments I've added should help clarify the reasoning for that promise callback mess.

File: src/store/Upgrade.test.ts [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+});
+
+describe("#newVersion", () => {
+  it("matches the version provided in the constructor", () => {
+    const oldVersion = 3;
+
+    const upgrade = new Upgrade(mockDB, oldVersion, 5, mockTX);
+
+    expect(upgrade.oldVersion).toEqual(oldVersion);
+  });
```
Comment (CJWaszczuk) [inline src/store/Upgrade.test.ts]: Should be newVersion

File: src/store/Upgrade.test.ts [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+
+    // Cast to get around TypeScript enforcing the type of the store name
+    // (not all users will be using TypeScript).
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const store = (upgrade as any).createStore(missingStoreName);
+
+    expect(store.name).toEqual(missingStoreName);
+  });
+});
+
```
Comment (CJWaszczuk) [inline src/store/Upgrade.test.ts]: Drop

File: src/store/Database.test.ts [hunk 7]
```diff
@@ -0,0 +1,679 @@
...
+    const currentVersion = 4;
+
+    initialDB = await Database.open<TestSchema>(testDBName, initialVersion);
+
+    initialDB.close();
+
+    await expectPromise(async () => {
+      db = await Database.open<TestSchema>(testDBName, currentVersion);
+    }).rejects.toThrow();
+  });
```
Comment (CJWaszczuk) [inline src/store/Database.test.ts]: Do we need to display an error here?

Comment (erbridge) [reply src/store/Database.test.ts]: Yes, we should be testing that. Thanks.

Comment (erbridge) [reply src/store/Database.test.ts]: Ah. The reason for this is that the error itself doesn't actually have a message. We might want to change that, but I don't think it's a priority.

**Pull request merged**